File: test-suite/golang/security/ssrf_clean.go
package security

import (
	"errors"
	"net/http"
	"net/url"
)

var allowedHosts = map[string]bool{
	"api.example.com":   true,
	"hooks.example.com": true,
}

func safeOutboundURL(raw string) (string, error) {
	parsed, err := url.Parse(raw)
	if err != nil {
		return "", err
	}
	if parsed.Scheme != "https" || !allowedHosts[parsed.Hostname()] {
		return "", errors.New("blocked outbound host")
	}
	return parsed.String(), nil
}

func proxyAllowedURL(r *http.Request) (*http.Response, error) {
	target, err := safeOutboundURL(r.URL.Query().Get("url"))
	if err != nil {
		return nil, err
	}
	return http.Get(target)
}

func postAllowedCallback(r *http.Request) (*http.Response, error) {
	target, err := safeOutboundURL(r.FormValue("callback"))
	if err != nil {
		return nil, err
	}
	return http.Post(target, "application/json", nil)
}

func fetchAllowedRequest(r *http.Request, client *http.Client) (*http.Response, error) {
	target, err := safeOutboundURL(r.Header.Get("X-Webhook-Url"))
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(r.Context(), http.MethodGet, target, nil)
	if err != nil {
		return nil, err
	}
	return client.Do(req)
}

File: test-suite/csharp/security/SsrfClean.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;

public sealed class SsrfClean
{
    private static readonly HashSet<string> AllowedHosts = new(StringComparer.OrdinalIgnoreCase)
    {
        "api.example.com",
        "hooks.example.com"
    };

    private readonly HttpClient _httpClient = new HttpClient();

    private static Uri SafeOutboundUrl(string raw)
    {
        if (!Uri.TryCreate(raw, UriKind.Absolute, out var uri) ||
            uri.Scheme != Uri.UriSchemeHttps ||
            !AllowedHosts.Contains(uri.Host))
        {
            throw new InvalidOperationException("blocked outbound URL");
        }

        return uri;
    }

    public Task<string> FetchQuery(HttpRequest request)
    {
        var target = SafeOutboundUrl(request.Query["url"]!);
        return _httpClient.GetStringAsync(target);
    }

    public Task<HttpResponseMessage> FetchCallback(HttpContext context)
    {
        var callback = SafeOutboundUrl(context.Request.Headers["X-Callback-Url"]!);
        return _httpClient.GetAsync(callback);
    }

    public Task<HttpResponseMessage> SendRouteEndpoint(HttpRequest request)
    {
        var endpoint = SafeOutboundUrl(request.RouteValues["endpoint"]?.ToString() ?? "https://api.example.com");
        var message = new HttpRequestMessage(HttpMethod.Get, endpoint);
        return _httpClient.SendAsync(message);
    }

    public WebRequest CreateWebhookRequest(HttpContext context)
    {
        var webhook = SafeOutboundUrl(context.Request.Form["webhookUrl"]!);
        return WebRequest.Create(webhook);
    }

    public string FetchHost(HttpRequest request)
    {
        var target = SafeOutboundUrl(request.Query["url"]!);
        using var webClient = new WebClient();
        return webClient.DownloadString(target);
    }

    public Task<HttpResponseMessage> FetchAfterInlineValidation(HttpRequest request)
    {
        var target = request.Query["nextUrl"].ToString();
        if (!Uri.TryCreate(target, UriKind.Absolute, out var uri) ||
            uri.Scheme != Uri.UriSchemeHttps ||
            !AllowedHosts.Contains(uri.Host))
        {
            return Task.FromResult(new HttpResponseMessage(HttpStatusCode.BadRequest));
        }

        return _httpClient.GetAsync(target);
    }
}
